File: uafricas_backend/src/models/admin/vidafrica.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

// ── Colonnes SQL ─────────────────────────────────────────────

pub const ADMIN_VIDEO_LISTE_COLONNES: &str =
    "v.id, v.titre, v.slug, v.vignette_url, v.duree_secondes, v.etat,
     (SELECT COUNT(*) FROM media_content.piste_sous_titre p
      WHERE p.video_id = v.id AND p.deleted_at IS NULL)::INTEGER AS nombre_pistes,
     v.created_at";

pub const ADMIN_VIDEO_DETAIL_COLONNES: &str =
    "v.id, v.titre, v.slug, v.description, v.fichier_video_url,
     v.vignette_url, v.duree_secondes, v.taille_octets, v.format_video,
     v.etat, v.cree_par, u.nom || ' ' || u.prenom AS cree_par_nom,
     v.created_at, v.updated_at";

pub const VIDEO_TRI_COLONNES: &[&str] = &[
    "created_at", "titre", "etat", "duree_secondes",
];

// ── Langues valides ──────────────────────────────────────────

pub const LANGUES_VALIDES: &[&str] = &[
    "francais", "anglais", "arabe", "portugais", "swahili",
    "wolof", "haoussa", "amharique", "zoulou", "lingala",
    "bambara", "yoruba", "peul", "espagnol", "mandarin",
];

// ── Labels langues ───────────────────────────────────────────

pub fn label_langue(code: &str) -> &str {
    match code {
        "francais" => "Français",
        "anglais" => "Anglais",
        "arabe" => "Arabe",
        "portugais" => "Portugais",
        "swahili" => "Swahili",
        "wolof" => "Wolof",
        "haoussa" => "Haoussa",
        "amharique" => "Amharique",
        "zoulou" => "Zoulou",
        "lingala" => "Lingala",
        "bambara" => "Bambara",
        "yoruba" => "Yoruba",
        "peul" => "Peul",
        "espagnol" => "Espagnol",
        "mandarin" => "Mandarin",
        _ => code,
    }
}

// ── Vidéos — Liste ───────────────────────────────────────────

#[derive(Debug, Serialize, FromRow)]
pub struct AdminVideoListeResponse {
    pub id: Uuid,
    pub titre: String,
    pub slug: String,
    pub vignette_url: Option<String>,
    pub duree_secondes: Option<i32>,
    pub etat: String,
    pub nombre_pistes: Option<i32>,
    pub created_at: DateTime<Utc>,
}

// ── Vidéos — Détail ──────────────────────────────────────────

#[derive(Debug, FromRow)]
pub struct AdminVideoDetailRow {
    pub id: Uuid,
    pub titre: String,
    pub slug: String,
    pub description: Option<String>,
    pub fichier_video_url: String,
    pub vignette_url: Option<String>,
    pub duree_secondes: Option<i32>,
    pub taille_octets: Option<i64>,
    pub format_video: Option<String>,
    pub etat: String,
    pub cree_par: Uuid,
    pub cree_par_nom: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct AdminVideoDetailResponse {
    pub id: Uuid,
    pub titre: String,
    pub slug: String,
    pub description: Option<String>,
    pub fichier_video_url: String,
    pub vignette_url: Option<String>,
    pub duree_secondes: Option<i32>,
    pub taille_octets: Option<i64>,
    pub format_video: Option<String>,
    pub etat: String,
    pub cree_par: Uuid,
    pub cree_par_nom: Option<String>,
    pub pistes: Vec<AdminPisteSousTitreResponse>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl AdminVideoDetailRow {
    pub fn to_response(&self, pistes: Vec<AdminPisteSousTitreResponse>) -> AdminVideoDetailResponse {
        AdminVideoDetailResponse {
            id: self.id,
            titre: self.titre.clone(),
            slug: self.slug.clone(),
            description: self.description.clone(),
            fichier_video_url: self.fichier_video_url.clone(),
            vignette_url: self.vignette_url.clone(),
            duree_secondes: self.duree_secondes,
            taille_octets: self.taille_octets,
            format_video: self.format_video.clone(),
            etat: self.etat.clone(),
            cree_par: self.cree_par,
            cree_par_nom: self.cree_par_nom.clone(),
            pistes,
            created_at: self.created_at,
            updated_at: self.updated_at,
        }
    }
}

// ── Query Params ─────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct AdminVideoQueryParams {
    pub page: Option<i64>,
    pub par_page: Option<i64>,
    pub tri_par: Option<String>,
    pub tri_dir: Option<String>,
    pub recherche: Option<String>,
    pub etat: Option<String>,
}

// ── Requêtes de mutation ─────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct ChangerEtatVideoRequest {
    pub etat: String,
}

#[derive(Debug, Deserialize)]
pub struct CreerPisteRequest {
    pub langue: String,
}

#[derive(Debug, Deserialize)]
pub struct CreerSegmentRequest {
    pub texte: String,
    pub debut_ms: i32,
    pub fin_ms: i32,
}

#[derive(Debug, Deserialize)]
pub struct ModifierSegmentRequest {
    pub texte: Option<String>,
    pub debut_ms: Option<i32>,
    pub fin_ms: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct ReordonnerSegmentsRequest {
    pub ordre: Vec<Uuid>,
}

#[derive(Debug, Deserialize)]
pub struct TimingMotItem {
    pub position: i32,
    pub mot: String,
    pub debut_ms: i32,
    pub fin_ms: i32,
}

#[derive(Debug, Deserialize)]
pub struct EnregistrerTimingsMotRequest {
    pub timings: Vec<TimingMotItem>,
}

// ── Pistes — Response ────────────────────────────────────────

#[derive(Debug, Serialize, FromRow)]
pub struct AdminPisteSousTitreResponse {
    pub id: Uuid,
    pub langue: String,
    pub est_complete: bool,
    pub nombre_segments: Option<i32>,
    pub created_at: DateTime<Utc>,
}

// ── Segments — Response ──────────────────────────────────────

#[derive(Debug, Serialize)]
pub struct AdminSegmentSousTitreResponse {
    pub id: Uuid,
    pub position: i32,
    pub texte: String,
    pub debut_ms: i32,
    pub fin_ms: i32,
    pub timings_mot: Vec<TimingMotResponse>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, FromRow)]
pub struct AdminSegmentRow {
    pub id: Uuid,
    pub position: i32,
    pub texte: String,
    pub debut_ms: i32,
    pub fin_ms: i32,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, FromRow)]
pub struct TimingMotResponse {
    pub position: i32,
    pub mot: String,
    pub debut_ms: i32,
    pub fin_ms: i32,
}

// ── Slug ─────────────────────────────────────────────────────

pub fn generer_slug(titre: &str) -> String {
    titre
        .to_lowercase()
        .replace(['é', 'è', 'ê', 'ë'], "e")
        .replace(['à', 'â', 'ä'], "a")
        .replace(['ù', 'û', 'ü'], "u")
        .replace(['î', 'ï'], "i")
        .replace(['ô', 'ö'], "o")
        .replace('ç', "c")
        .replace(|c: char| !c.is_alphanumeric() && c != ' ', "")
        .split_whitespace()
        .collect::<Vec<&str>>()
        .join("-")
}

File: uafricas_frontend/app/mocks/vidafrica.ts
// ── Interfaces ───────────────────────────────────────────────

export type LangueSousTitre =
  | 'francais' | 'anglais' | 'arabe' | 'portugais' | 'swahili'
  | 'wolof' | 'haoussa' | 'amharique' | 'zoulou' | 'lingala'
  | 'bambara' | 'yoruba' | 'peul' | 'espagnol' | 'mandarin'

export const LANGUES_LABELS: Record<string, string> = {
  francais: 'Français',
  anglais: 'Anglais',
  arabe: 'Arabe',
  portugais: 'Portugais',
  swahili: 'Swahili',
  wolof: 'Wolof',
  haoussa: 'Haoussa',
  amharique: 'Amharique',
  zoulou: 'Zoulou',
  lingala: 'Lingala',
  bambara: 'Bambara',
  yoruba: 'Yoruba',
  peul: 'Peul',
  espagnol: 'Espagnol',
  mandarin: 'Mandarin',
}

export interface TimingMot {
  position: number
  mot: string
  debut_ms: number
  fin_ms: number
}

export interface SegmentSousTitre {
  id: string
  position: number
  texte: string
  debut_ms: number
  fin_ms: number
  timings_mot: TimingMot[]
  created_at: string
  updated_at: string
}

export interface PisteSousTitre {
  id: string
  langue: LangueSousTitre
  est_complete: boolean
  etat?: string
  cree_par?: string
  cree_par_nom?: string | null
  nombre_segments: number
  created_at: string
}

export interface Video {
  id: string
  titre: string
  slug: string
  description: string | null
  fichier_video_url: string
  vignette_url: string | null
  duree_secondes: number | null
  taille_octets: number | null
  format_video: string | null
  etat: string
  cree_par: string
  cree_par_nom: string | null
  pistes: PisteSousTitre[]
  created_at: string
  updated_at: string
}

export interface VideoListe {
  id: string
  titre: string
  slug: string
  vignette_url: string | null
  duree_secondes: number | null
  etat: string
  nombre_pistes: number
  created_at: string
}

// ── Données mock ─────────────────────────────────────────────

export const mockVideos: VideoListe[] = [
  {
    id: 'v1-uuid',
    titre: 'Introduction à la culture panafricaine',
    slug: 'introduction-a-la-culture-panafricaine',
    vignette_url: null,
    duree_secondes: 320,
    etat: 'publie',
    nombre_pistes: 2,
    created_at: '2026-04-10T10:00:00Z',
  },
  {
    id: 'v2-uuid',
    titre: 'Les langues africaines : richesse et diversité',
    slug: 'les-langues-africaines-richesse-et-diversite',
    vignette_url: null,
    duree_secondes: 540,
    etat: 'brouillon',
    nombre_pistes: 0,
    created_at: '2026-04-11T14:30:00Z',
  },
  {
    id: 'v3-uuid',
    titre: 'Documentaire : la musique traditionnelle',
    slug: 'documentaire-la-musique-traditionnelle',
    vignette_url: null,
    duree_secondes: 1200,
    etat: 'publie',
    nombre_pistes: 3,
    created_at: '2026-04-12T09:15:00Z',
  },
]

export const mockPistes: PisteSousTitre[] = [
  {
    id: 'p1-uuid',
    langue: 'francais',
    est_complete: true,
    nombre_segments: 3,
    created_at: '2026-04-10T11:00:00Z',
  },
  {
    id: 'p2-uuid',
    langue: 'anglais',
    est_complete: false,
    nombre_segments: 2,
    created_at: '2026-04-10T12:00:00Z',
  },
]

export const mockSegments: SegmentSousTitre[] = [
  {
    id: 's1-uuid',
    position: 1,
    texte: 'Bienvenue dans ce documentaire',
    debut_ms: 1000,
    fin_ms: 3500,
    timings_mot: [
      { position: 1, mot: 'Bienvenue', debut_ms: 1000, fin_ms: 1800 },
      { position: 2, mot: 'dans', debut_ms: 1800, fin_ms: 2100 },
      { position: 3, mot: 'ce', debut_ms: 2100, fin_ms: 2400 },
      { position: 4, mot: 'documentaire', debut_ms: 2400, fin_ms: 3500 },
    ],
    created_at: '2026-04-10T11:30:00Z',
    updated_at: '2026-04-10T11:30:00Z',
  },
  {
    id: 's2-uuid',
    position: 2,
    texte: 'Nous allons explorer ensemble',
    debut_ms: 4000,
    fin_ms: 6500,
    timings_mot: [
      { position: 1, mot: 'Nous', debut_ms: 4000, fin_ms: 4500 },
      { position: 2, mot: 'allons', debut_ms: 4500, fin_ms: 5000 },
      { position: 3, mot: 'explorer', debut_ms: 5000, fin_ms: 5800 },
      { position: 4, mot: 'ensemble', debut_ms: 5800, fin_ms: 6500 },
    ],
    created_at: '2026-04-10T11:35:00Z',
    updated_at: '2026-04-10T11:35:00Z',
  },
]

// ── Helpers ──────────────────────────────────────────────────

const delay = (ms: number = 300) => new Promise(resolve => setTimeout(resolve, ms))

export const getVideoParId = async (id: string): Promise<Video | null> => {
  await delay()
  const v = mockVideos.find(v => v.id === id)
  if (!v) return null
  return {
    ...v,
    description: 'Description de la vidéo mock',
    fichier_video_url: '/uploads/videos/sample.mp4',
    taille_octets: 52428800,
    format_video: 'mp4',
    cree_par: 'admin-uuid',
    cree_par_nom: 'Admin Test',
    pistes: mockPistes,
    updated_at: v.created_at,
  }
}

export const getVideoParSlug = async (slug: string): Promise<Video | null> => {
  await delay()
  const v = mockVideos.find(v => v.slug === slug)
  if (!v) return null
  return {
    ...v,
    description: 'Description de la vidéo mock',
    fichier_video_url: '/uploads/videos/sample.mp4',
    taille_octets: 52428800,
    format_video: 'mp4',
    cree_par: 'admin-uuid',
    cree_par_nom: 'Admin Test',
    pistes: mockPistes,
    updated_at: v.created_at,
  }
}

export const filtrerVideos = async (params: {
  recherche?: string
  etat?: string
}): Promise<VideoListe[]> => {
  await delay()
  return mockVideos.filter((v) => {
    if (params.etat && v.etat !== params.etat) return false
    if (params.recherche) {
      const r = params.recherche.toLowerCase()
      if (!v.titre.toLowerCase().includes(r)) return false
    }
    return true
  })
}

export const formaterDuree = (secondes: number | null): string => {
  if (!secondes) return '--:--'
  const min = Math.floor(secondes / 60)
  const sec = secondes % 60
  return `${min}:${sec.toString().padStart(2, '0')}`
}

export const formaterTimestamp = (ms: number): string => {
  const sec = Math.floor(ms / 1000)
  const min = Math.floor(sec / 60)
  const reste = sec % 60
  const millis = ms % 1000
  return `${min}:${reste.toString().padStart(2, '0')}.${millis.toString().padStart(3, '0')}`
}
